backends/pkg/reconciler/eventmesh/handler.go

```diff
@@ -32,63 +33,81 @@ func EventMeshHandler(ctx context.Context, listers Listers) func(w http.Response
-
-		convertedEventTypeMap := make(EventTypeMap)
-		for _, et := range fetchedEventTypes {
-			namespaceEventTypeRef := NamespaceEventTypeRef(et)
+func BuildEventMesh(listers Listers, logger *zap.SugaredLogger) (error, EventMesh) {
+	err, convertedBrokers := fetchBrokers(listers.BrokerLister, logger)
+	if err != nil {
+		logger.Errorw("Error fetching and converting brokers", "error", err)
+		return err, EventMesh{}
+	}
 
-			if et.Spec.Reference != nil {
-				if br, ok := brokerMap[RefNameAndNamespace(et.Spec.Reference)]; ok {
-					// add to broker provided event types
-					// only add if it hasn't been added already
-					if !slices.Contains(br.ProvidedEventTypes, namespaceEventTypeRef) {
-						br.ProvidedEventTypes = append(br.ProvidedEventTypes, namespaceEventTypeRef)
-					}
-				}
-			}
+	brokerMap := make(map[string]*Broker)
+	for _, cbr := range convertedBrokers {
+		brokerMap[cbr.GetNameAndNamespace()] = cbr
+	}
 
-			if _, ok := convertedEventTypeMap[namespaceEventTypeRef]; ok {
-				logger.Debugw("Duplicate event type", "event type", namespaceEventTypeRef)
-				continue
-			}
+	fetchedEventTypes, err := listers.EventTypeLister.List(labels.Everything())
+	if err != nil {
+		logger.Errorw("Error listing eventTypes", "error", err)
+		return err, EventMesh{}
+	}
 
-			convertedEventType := convertEventType(et)
-			convertedEventTypeMap[namespaceEventTypeRef] = &convertedEventType
+	sort.Slice(fetchedEventTypes, func(i, j int) bool {
+		if fetchedEventTypes[i].Namespace != fetchedEventTypes[j].Namespace {
+			return fetchedEventTypes[i].Namespace < fetchedEventTypes[j].Namespace
 		}
+		return fetchedEventTypes[i].Name < fetchedEventTypes[j].Name
+	})
 
-		eventMesh := EventMesh{
-			EventTypes: make([]*EventType, 0, len(convertedEventTypeMap)),
-			Brokers:    convertedBrokers,
-		}
+	logger.Debugw("Fetched event types", "event types", fetchedEventTypes)
 
-		for _, et := range convertedEventTypeMap {
-			eventMesh.EventTypes = append(eventMesh.EventTypes, et)
+	convertedEventTypeMap := make(EventTypeMap)
+	for _, et := range fetchedEventTypes {
+		namespaceEventTypeRef := NamespaceEventTypeRef(et)
+
+		if et.Spec.Reference != nil {
+			if br, ok := brokerMap[RefNameAndNamespace(et.Spec.Reference)]; ok {
+				// add to broker provided event types
+				// only add if it hasn't been added already
+				if !slices.Contains(br.ProvidedEventTypes, namespaceEventTypeRef) {
+					br.ProvidedEventTypes = append(br.ProvidedEventTypes, namespaceEventTypeRef)
+				}
+			}
 		}
 
-		err = json.NewEncoder(w).Encode(eventMesh)
-		if err != nil {
-			logger.Errorw("Error encoding event mesh", "error", err)
-			return
+		if _, ok := convertedEventTypeMap[namespaceEventTypeRef]; ok {
+			logger.Debugw("Duplicate event type", "event type", namespaceEventTypeRef)
+			continue
 		}
+
+		convertedEventType := convertEventType(et)
+		convertedEventTypeMap[namespaceEventTypeRef] = &convertedEventType
```

Comment:
If the EventType has a different source but same type, we would show only one (at random), why is that the case ?

Reply:
Unless @matzew has a better idea, I don't want to show all duplicate event types in Backstage. Creates lots of confusion. In Backstage, the important information is what event types are available to subscribe to and which brokers give them.

About random: yes, I couldn't find a better determinism other than using the information of the first one with the same `spec.type`.

Reply:
I believe it's confusing because we're using the type as the displayed name but it doesn't have to be like that, we can use a combination of attributes.

> In Backstage, the important information is what event types are available to subscribe to and which brokers give them.

I think not only brokers, sources as well, right ?

Reply:
is there a restriction on the character set for the displayed name ?

But until then I wouldn't deduplicate eventtypes as they aren't duplicate, they represent different information

Reply:
> is there a restriction on the character set for the displayed name ?

What's on your mind to show on Backstage?

Let me give some examples.

## Current look:

![Screenshot 2023-12-12 at 13 01 16](https://github.com/knative-extensions/backstage-plugins/assets/376732/57f7ac01-97dd-4497-b4e7-eb8ec9f989ae)

## No deduplication - showing `spec.type` as the title

![Screenshot 2023-12-12 at 13 08 44](https://github.com/knative-extensions/backstage-plugins/assets/376732/318fbe17-0e10-46c8-a929-c5da62b3168f)

## No deduplication - showing `name` as the title

![Screenshot 2023-12-12 at 13 08 13](https://github.com/knative-extensions/backstage-plugins/assets/376732/95b6818d-88b1-4f3e-ac97-bb32358408fc)

## No deduplication - showing more stuff (longer titles)

![Screenshot 2023-12-12 at 13 09 38](https://github.com/knative-extensions/backstage-plugins/assets/376732/7a9dc4ac-bd72-4c59-832b-50284688e714)

Reply:
The problem here is that the underlying mechanism of event types are creating duplicates. And we shouldn't reflect the same information on Backstage as is in my opinion.

Reply:
> they represent different information

We need to evaluate that if those different information is really important for users of Backstage. I don't know yet.

Reply:
> The problem here is that the underlying mechanism of event types are creating duplicates. And we shouldn't reflect the same information on Backstage as is in my opinion.

They are not, they represent different sources

Reply:
Nice then, I would show the source as well, maybe in parenthesis or some other format
```
dev.knative.eventing.resource.add (<source>)
```

Reply:
You mean the `spec.Reference`? Or actually `spec.source`?

Neither of these provide uniqueness. 
Uniqueness is based on a tuple of more properties AFAI understand. (not sure 100% what the props are anymore).

And using all those information in the title could be problematic.

Reply:
> I don't want to show all duplicate event types in Backstage. Creates lots of confusion.

Imagine two cloud events, each have the same `type` set to same value, **BUT** have a different source attribute (or any other attribute difference), those are two entities we want to show.


Please check the two related feature tracks for this: 
* https://github.com/knative/eventing/issues/7264
* https://github.com/knative/eventing/issues/7265

they might provide some extra info.

But we can not ignore Events being shown, when their attributes differ in just one field. If so, they are not duplicates

Reply:
So, imo these are two different objects/events:

```
  specversion: 1.0
  type: dev.knative.sources.ping
  source: /apis/v1/namespaces/pingsource-example/pingsources/first-ping-source
```

and 
```
  specversion: 1.0
  type: dev.knative.sources.ping
  source: /apis/v1/namespaces/pingsource-example/pingsources/second-ping-source
```

`ID` and `time` can be safely ignored, but substantical attributes, like `source` should be reflected.

Reply:
When we allow to have EventTypes to live with NO `reference`, they are considered more global/abstract (remember that discussion, I know we talked about that before), and can be shown even there is no running eventing workload on the cluster.

and when they have a refernece they are more for actual running workloads.

Perhaps we can difrerence there?  But this is not yet implemented...

Reply:
Backstage is not the ultimate Developer Console. It will show the overview of having events flowing from A to B.

I don't see too much value of showing 5 events in Backstage that have the same type.

However, I will fallback to not deduplicating anything for now.

Reply:
It will look like this:

![Screenshot 2023-12-18 at 15 54 42](https://github.com/knative-extensions/backstage-plugins/assets/376732/88c773ea-f78b-439a-b10c-6ade16591a7a)

Reply:
> Backstage is not the ultimate Developer Console. It will show the overview of having events flowing from A to B.

I think I see what you mean.

## generic discovery

How about we focus initially on those events that are having _no_ `reference` set, as we will treat them as a generic thing, as discussed in here: https://docs.google.com/document/d/1US5Ve0CwhXwO8pUudN36vLkuSetEtSgwQ1btuG_9Xmk/edit#heading=h.n8a530nnrb

For these cases there is likely less of a need to even duplicate, b/c for `ApiServerSource` there will be exact six EventTypes that are there in the cluster w/o any `reference`.  For `PingSource` there will be exactly one.

Would that simplify things for a more generic "event catalog" ? 


## actual workloads

Those that have a `reference` present, in a given namespace, are "live" event types, from existing workloads. Here (e.g. for creating triggers) it is more important to actually differentiate on the  diff of values (see my example with two `PingSource` workloads